File: python/app/judge.py
from time import perf_counter
from typesafe_sdk import AsyncTypeSafeClient, Choice, RetryPolicy
from app.game import CONFIG, character, vote_for


async def judge(data, character_id):
    person = character(character_id)
    state = dict(pitch=data["pitch"], previousPitch=data["previousPitch"])
    started = perf_counter()
    async with AsyncTypeSafeClient(
        retry=RetryPolicy(max_retries=0, timeout=30)
    ) as client:
        result = await client.system_one(
            state=state,
            questions={
                c["id"]: Choice(
                    instructions=f"{CONFIG['instructions']}\nCriterion: {c['question']}",
                    criteria=CONFIG["options"],
                )
                for c in person["criteria"]
            },
        )
    judgments = []
    for criterion in person["criteria"]:
        answer = result.choices[criterion["id"]]
        if answer.choice not in CONFIG["options"]:
            raise ValueError("Missing TypeSafe criterion result.")
        judgments.append(
            {
                **criterion,
                "choice": answer.choice,
                "confidence": answer.confidence,
                "probabilities": answer.probabilities,
            }
        )
    return dict(
        stage="character",
        characterId=character_id,
        name=person["name"],
        vote=vote_for(judgments),
        previousVote=next(
            (r["vote"] for r in data["previous"] if r["characterId"] == character_id),
            None,
        ),
        judgments=judgments,
        typesafeModel=result.model,
        decisionMs=round((perf_counter() - started) * 1000),
        rulesVersion=CONFIG["rulesVersion"],
        instructions=CONFIG["instructions"],
        options=CONFIG["options"],
        state=state,
        minimumConfidence=CONFIG["minimumConfidence"],
    )

File: typescript/src/judge.ts
import { choice, TypeSafeClient } from "@typesafe-ai/sdk";
import config from "../../shared/config.json";
import { character, voteFor } from "./game";
import type {
  CharacterResult,
  Finding,
  Judgment,
  RoundInput,
} from "../../shared/types";

export async function judge(
  input: RoundInput,
  characterId: string,
): Promise<
  Omit<
    CharacterResult,
    "reaction" | "reactionSource" | "dialogueModel" | "durationMs"
  >
> {
  const person = character(characterId);
  const state = { pitch: input.pitch, previousPitch: input.previousPitch };
  const started = performance.now();
  const client = new TypeSafeClient({
    timeout: 30_000,
    retry: { maxRetries: 0 },
  });
  const result = await client.systemOne({
    state,
    questions: Object.fromEntries(
      person.criteria.map((c) => [
        c.id,
        choice(
          `${config.instructions}\nCriterion: ${c.question}`,
          config.options,
        ),
      ]),
    ),
  });
  const judgments: Judgment[] = person.criteria.map((c) => {
    const answer = result.answers[c.id];
    if (!answer || !(answer.choice in config.options))
      throw new Error("Missing TypeSafe criterion result.");
    return {
      ...c,
      choice: answer.choice as Finding,
      confidence: answer.confidence,
      probabilities: answer.probabilities as Record<Finding, number>,
    };
  });
  return {
    stage: "character",
    characterId,
    name: person.name,
    vote: voteFor(judgments),
    previousVote:
      input.previous.find((r) => r.characterId === characterId)?.vote ?? null,
    judgments,
    typesafeModel: result.model,
    decisionMs: Math.round(performance.now() - started),
    rulesVersion: config.rulesVersion,
    instructions: config.instructions,
    options: config.options,
    state,
    minimumConfidence: config.minimumConfidence,
  };
}
